Make sure that 'tox -erust' fails on bad RC.

kevinhartman opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Previously, a non-zero exit code from the subprocess would not fail the tox run like we'd expect.

kevinhartman committed:
Make sure that 'tox -erust' fails on bad RC.

Previously, a non-zero exit code from the subprocess would not
fail the tox run like we'd expect.

mtreinish commented on tox.ini:
```diff
@@ -41,7 +41,7 @@ commands =
   import os, subprocess, sysconfig;\
   os.environ["LD_LIBRARY_PATH"] = os.pathsep.join(filter(None, [sysconfig.get_config_var("LIBDIR"), os.getenv("LD_LIBRARY_PATH")]));\
-  subprocess.run(["cargo", "test", "--no-default-features"])'
+  subprocess.run(["cargo", "test", "--no-default-features"]).check_returncode()'
```
Not that it particularly matters in this context. But what I normally do is:

```suggestion
  subprocess.run(["cargo", "test", "--no-default-features"], check=True)'
```

jakelishman replied:
I think pylint has a lint against failing to check the return code of `subprocess.run` - it might be a motivation to move this inline script into a proper file in the `tools` directory so it's subject to lint?

jakelishman replied:
It does: https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/subprocess-run-check.html

kevinhartman replied:
Seems like a good idea, and should let us use the same script from tox and CI.

Done in 663eaa8.

mtreinish reviewed (dismissed):
LGTM, just left a small comment inline. It doesn't really matter in this context though.

kevinhartman committed:
Use script to run cargo test, simplify subprocess call.

mtreinish commented on tools/run_cargo_test.py:
```diff
@@ -0,0 +1,40 @@
+
+# On Linux, the test executable's RPATH doesn't contain libpython, so we add it
+# to the dlopen search path here.
+os.environ["LD_LIBRARY_PATH"] = os.pathsep.join(
+    filter(None, [sysconfig.get_config_var("LIBDIR"), os.getenv("LD_LIBRARY_PATH")])
+)
```
Is there a reason to not just do?:
```suggestion
os.environ["LD_LIBRARY_PATH"] = os.pathsep.join([
    sysconfig.get_config_var("LIBDIR"),
    os.getenv("LD_LIBRARY_PATH")
])
```

I've always found the `filter(None, ...)` pattern a bit odd when you can just use a generator expression.

kevinhartman replied:
The reason I did that was to filter out `None` in the case that `"LD_LIBRARY_PATH"` isn't actually set to anything. I can do something different if you prefer!

mtreinish reviewed (commented):
This LGTM for the most part. I like having the script in a dedicated file now. Just one question/suggestion inline.

kevinhartman committed:
Use PYTHONPATH instead of PYTHONUSERBASE.

This fixes an issue where the site-packages added to sys.path
could be wrong (e.g. if using Framework-based Python on macOS).

This was happening because the venv would install qiskit using
one pathing scheme while the Python interpreter baked into the test
executable would use (potentially) another when generating the path
from PYTHONUSERBASE.

See https://docs.python.org/3/library/sysconfig.html#installation-paths
for details.

To resolve this, we just set the path to the venv's site-packages via
PYTHONPATH.

kevinhartman committed:
Add CWD to search path when running run_cargo_test.py

kevinhartman committed:
Update docs.

kevinhartman commented:
While testing this on his machine, @raynelfss discovered an issue with using the existing approach with `PYTHONUSERBASE` to get the virtual environment's `site-packages` into the embedded Python interpreter's path. The issue was that `venvs` use a potentially different folder layout when installing packages than the system's Python will use when looking for them, so the path returned by [site.USER_SITE](https://docs.python.org/3/library/site.html#site.USER_SITE) was incorrect.

Instead, we now set the `PYTHONPATH` to directly be the actual full path to the active Python interpreter's `site-packages`, which is more natural anyways. This wasn't done originally because before this PR, we weren't actually invoking `cargo test` from within a Python script.

Anyway, this should be good to go now.

jakelishman reviewed (approved):
Given that Matt's question was minor, and from offline conversations, this is important for people, let's merge now.  If there's more to tweak, we can follow up.
